### PR TITLE
HybridBlock model for Biaffine model

### PR DESCRIPTION
**Pull request by szha**
Hey Han,

As promised, here's a code sample for a hybrid-block model that utilizes fused kernel for RNN. I've only tested it on my laptop so far. The change roughly translates to 3x speed up on my laptop.

A couple more things to consider:
- there are several transposes in the code base, some of which may not be necessary. Since MXNet doesn't yet support column-major layout, these transposes, when done in mxnet, results in data shuffling which might take quite some time.
- the bilinear kernel currently requires batch size, which prevents that part of code from being hybridized. there should be ways around it and I'm happy to explore with you if you'd like to look into it.

### CODE REVIEW

**Comment by hankcs** [comment]
Thank you! I'll look into it after school.

**Comment by hankcs** [comment]
Hi Shen,

I confirmed a 3x speedup, but this new model seems not converging, after 100 steps, it gives:

```
git checkout master
python3 train.py --config configs/ptb.ini

Step #99: Acc: arc 0.03, rel 0.00, overall 0.00, loss 3.746
duration : 28.14
Test on development set
LAS 0.27, UAS 1.24
```

For my naive implementation, although not accurately implemented, but it does converge.

```
git checkout naive
python3 train.py --config configs/ptb.ini

Step #99: Acc: arc 0.64, rel 0.83, overall 0.57, loss 0.838

Test on development set
LAS 1.56, UAS 55.85
```

File: train.py [hunk 1]
```diff
@@ -31,15 +34,14 @@
...
+    with (mx.gpu(0) if 'cuda' in os.environ['PATH'] else mx.cpu()):
         parser = BiaffineParser(vocab, config.word_dims, config.tag_dims, config.dropout_emb, config.lstm_layers,
                             config.lstm_hiddens, config.dropout_lstm_input, config.dropout_lstm_hidden,
                             config.mlp_arc_size,
                             config.mlp_rel_size, config.dropout_mlp, config.debug)
-        parser.initialize(force_reinit=True)
         data_loader = DataLoader(config.train_file, config.num_buckets_train, vocab)
         # trainer = dy.AdamTrainer(pc, config.learning_rate, config.beta_1, config.beta_2, config.epsilon)
-        trainer = gluon.Trainer(parser.collect_params(), 'adam', {'learning_rate': config.learning_rate})
+        trainer = gluon.Trainer(parser.model.collect_params(), 'adam', {'learning_rate': config.learning_rate})
```
Comment (szha) [inline train.py:44]: seems that this should have been parser.collect_params(), because otherwise the weights in the outer model is not updating.

Comment (hankcs) [reply train.py:44]: Yes, that's true. Now it's good:

```
Step #99: Acc: arc 0.67, rel 0.88, overall 0.62, loss 0.758
duration : 28.53
Test on development set
LAS 1.44, UAS 63.82
```
